ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -0,0 +1,137 @@
+import com.github.ambry.quota.QuotaRecommendation;
+import com.github.ambry.quota.QuotaResource;
+import com.github.ambry.quota.QuotaSource;
+import com.github.ambry.quota.QuotaUtils;
+import com.github.ambry.rest.RestRequest;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Implementation of {@link QuotaEnforcer} for CU quota of Ambry's {@link QuotaResource}s.
+ * Note that this implementation is not thread safe or atomic. The caller should take care of thread safety guarantees
+ * where needed.
+ */
+public class AmbryCUQuotaEnforcer implements QuotaEnforcer {
+  private final static List<QuotaName> SUPPORTED_QUOTA_NAMES =
+      Collections.unmodifiableList(Arrays.asList(QuotaName.READ_CAPACITY_UNIT, QuotaName.WRITE_CAPACITY_UNIT));
+  private final static long NO_THROTTLE_RETRY_AFTER_MS = -1;
+  private final QuotaSource quotaSource;
+  private final float maxFrontendCuUsageToAllowExceed;
+  private final long throttleRetryAfterMs;
+
+  /**
+   * Constructor for {@link AmbryCapacityUnitQuotaEnforcer}.
+   * @param quotaSource {@link QuotaSource} where the quota limit and usage will be saved and retrieved from.
+   */
+  public AmbryCUQuotaEnforcer(QuotaSource quotaSource, QuotaConfig quotaConfig) {
+    this.quotaSource = quotaSource;
+    this.throttleRetryAfterMs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    this.maxFrontendCuUsageToAllowExceed = quotaConfig.maxFrontendCuUsageToAllowExceed;
+  }
+
+  @Override
+  public void init() {
+  }
+
+  @Override
+  public void charge(RestRequest restRequest, BlobInfo blobInfo, Map<QuotaName, Double> requestCostMap)
+      throws QuotaException {
+    final QuotaName quotaName = QuotaUtils.getCUQuotaName(restRequest);
+    try {
+      if (requestCostMap.size() == 0 || !requestCostMap.containsKey(quotaName)) {
+        String errorMessage = String.format("No %s cost provided for request %s. Nothing to charge", quotaName.name(),
+            QuotaUtils.requestDetailsToString(restRequest, blobInfo));
+        throw new QuotaException(errorMessage, true);
+      }
+      final QuotaResource quotaResource = QuotaResource.fromRestRequest(restRequest);
+      quotaSource.chargeUsage(quotaResource, quotaName, requestCostMap.get(quotaName));
+      quotaSource.chargeSystemResourceUsage(quotaName, requestCostMap.get(quotaName));
+    } catch (QuotaException qEx) {
+      throw new QuotaException(String.format("Could not charge for request %s due to %s",
+          QuotaUtils.requestDetailsToString(restRequest, blobInfo), qEx.getMessage()), qEx.isRetryable());
+    } catch (Exception ex) {
+      throw new QuotaException(String.format("Could not charge for request %s due to unexpected exception %s",
+          QuotaUtils.requestDetailsToString(restRequest, blobInfo), ex.getMessage()), true);
+    }
+  }
+
+  @Override
+  public QuotaRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    try {
```

Comment:
I think whatever checks you are making in charge(), you should also make them here. Otherwise, you may run into exceptions in charge() after passing the recommend(). That's bad if you consistently fail during the second enforcer as you will keep charging for the first one.

Reply:
We should have a metric here to set up alerts on, so that we catch the situation of over charging system resource quickly.